Add memory format support to `empty_like` operator

VitalyFedyunin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28544 Add torch.memory_format support to the TorchScript
* #28292 Add memory format support to the `resize_` op.
* #28291 TensroIterator preserve format for binary, ternary operators.
* #28077 Add channels last support to cuda.comm.scatter and gather
* #28076 Kill `operator==` of TensorOptions as confusing one
* #27979 Add memory format support to `resize_as_` operator
* #27891 Cleanup testing of _like operators
* #27890 Add memory format support to `randn_like` operator
* #27889 Add memory format support to `randint_like` operator
* #27562 Add memory format support to `zeros_like` operator
* #27561 Add memory format support to `rand_like` operator
* #27270 Add memory format support to `ones_like` operator
* #27262 Add memory format support to `full_like` operator
* **#27244 Add memory format support to `empty_like` operator**

Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980310](https://our.internmc.facebook.com/intern/diff/D17980310)

VitalyFedyunin committed:
Add memory format support to `empty_like` operator

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Add memory format support to `empty_like` operator"




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Add memory format support to `empty_like` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Add memory format support to `empty_like` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Add memory format support to `empty_like` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Add memory format support to `empty_like` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Fix merge conflict on "Add memory format support to `empty_like` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Fix .to copy=True path on "Add memory format support to `empty_like` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Rebase on top of master on "Add memory format support to `empty_like` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Fixing casting ops and tests on "Add memory format support to `empty_like` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Rebasing to master on "Add memory format support to `empty_like` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Rebasing to master on "Add memory format support to `empty_like` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980310](https://our.internmc.facebook.com/intern/diff/D17980310)

[ghstack-poisoned]

VitalyFedyunin commented:
@houseroad I had to whitelist `empty_like` inside of check_backward_compatibility.py as it doen't like changing of default value 
```
processing existing schema:  aten::empty_like.dtype(Tensor self, *, int dtype, int layout, Device device, bool pin_memory=False, int? memory_format=0) -> (Tensor)
Can NOT find backward compatible schemas after changes for schema aten::empty_like.dtype(Tensor self, *, int dtype, int layout, Device device, bool pin_memory=False, int? memory_format=0) -> (Tensor) from the following candidates:
[
aten::empty_like(Tensor self, *, int? memory_format=None) -> (Tensor)
	aten::empty_like.dtype(Tensor self, *, int dtype, int layout, Device device, bool pin_memory=False, int? memory_format=None) -> (Tensor)
]
```

houseroad commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1115,12 +1115,11 @@
   supports_named_tensor: True
 
-- func: empty_like.dtype(Tensor self, *, ScalarType dtype, Layout layout, Device device, bool pin_memory=False, MemoryFormat? memory_format=contiguous_format) -> Tensor
+- func: empty_like.dtype(Tensor self, *, ScalarType dtype, Layout layout, Device device, bool pin_memory=False, MemoryFormat? memory_format=None) -> Tensor
```
None and contigous_format (i.e., 0) have the same behavior?

VitalyFedyunin replied:
Yes, but it is exceptional case. No need to handle it in the test.

houseroad replied:
Sounds fine.

VitalyFedyunin commented:
@ailzhang hi

ezyang commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -196,22 +198,19 @@ Tensor empty_like(
+  if (self.is_quantized()) {
+
+    auto memory_format =
+        optional_memory_format.value_or(MemoryFormat::Contiguous);
```
To be changed later right :>

VitalyFedyunin replied:
Yes, right now we need to introduce this option, after we check all current code for expected behaviour and change default.

zou3519 replied:
If it's to be changed later, can we add a comment linking some issue or with some information ?

ezyang replied:
What Richard said

ezyang commented on tools/autograd/gen_variable_factories.py:
```diff
@@ -64,7 +64,10 @@ def process_function(decl, has_tensor_options, disable_autograd):
     if decl['name'].endswith('_like') and not has_tensor_options:
         # it's a tensor
-        actuals.append('{}.options().is_variable(false)'.format(actuals[0]))
+        if decl['name'] == 'empty_like':
```
Umm... this looks very non-general. (I assume it gets fixed later in the stack but it is very shocking here)

VitalyFedyunin replied:
Yes, need to list by function until I convert all of them

ezyang commented on test/test_torch.py:
```diff
@@ -12371,7 +12371,7 @@ def check_sum_all(tensor):
             check_sum_all(torch.tensor([True, False, True], dtype=torch.bool, device=device))
 
-    def _test_memory_format_transformations(self, device, input_generator_fn, transformation_fn):
+    def _test_memory_format_transformations(self, device, input_generator_fn, transformation_fn, compare_data=True):
```
Goofy but ok.

VitalyFedyunin committed:
Fix rand_like test on "Add memory format support to `empty_like` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980310](https://our.internmc.facebook.com/intern/diff/D17980310)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix resize_as_ linter on "Add memory format support to `empty_like` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980310](https://our.internmc.facebook.com/intern/diff/D17980310)

[ghstack-poisoned]

VitalyFedyunin committed:
Rebase on "Add memory format support to `empty_like` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980310](https://our.internmc.facebook.com/intern/diff/D17980310)

[ghstack-poisoned]

VitalyFedyunin committed:
randn_like not simple mapable any more on "Add memory format support to `empty_like` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980310](https://our.internmc.facebook.com/intern/diff/D17980310)

[ghstack-poisoned]

VitalyFedyunin committed:
fix bc test for empty_like on "Add memory format support to `empty_like` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980310](https://our.internmc.facebook.com/intern/diff/D17980310)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1116,12 +1116,11 @@
 
-- func: empty_like(Tensor self) -> Tensor
-  use_c10_dispatcher: full
+- func: empty_like(Tensor self, *, MemoryFormat? memory_format=None) -> Tensor
```
With the benefit of hindsight, I should have asked: why do we have two overloads of `empty_like` in the first place?!

VitalyFedyunin replied:
Do you mean this line and line 1123?

ezyang replied:
yes. But see my most recent PR.

VitalyFedyunin replied:
As far as I remember combining them together was never considered by me.
